### PR TITLE
CET supported by default

### LINKED ISSUE #42600: [Breaking change]: .NET 9 enables CET support by default
### Description

In .NET 9 Preview 6, the apphost and singlefilehost started to be marked as Intel CET compatible (they are compiled with /CETCOMPAT option). It was done to enhance security of .NET applications. That imposes a limitation on shared libraries that .NET applications can load and interop with. They are not allowed to set thread context to a location with instruction pointer that is not present on the shadow stack or in a table of allowed continuation addresses for exception handling.

### Version

Other (please put exact version in description textbox)

### Previous behavior

Shared libraries loaded into .NET process were able to set thread context using SetThreadContext, RtlRestoreContext/NtContinue or using their exception handlers to any location in the process address space.

### New behavior

Shared libraries loaded into .NET process are only allowed to set thread context using SetThreadContext, RtlRestoreContext/NtContinue or using their exception handlers to locations that are present on the shadow stack or in a table of allowed continuation addresses for exception handling (generated by /EHCONT compiler option or the `SetProcessDynamicEHContinuationTargets` API). Trying to change a thread context to any other location results in termination of the process. 

### Type of breaking change

- [X] **Binary incompatible**: Existing binaries might encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code might require source changes to compile successfully.
- [ ] **Behavioral change**: Existing binaries might behave differently at run time.

### Reason for change

Enabling CET for .NET applications enhance the security of the .NET applications by adding hardware-enforced stack protection that offers robust protection against ROP exploits (return oriented programming).

### Recommended action

Workarounds:
* Adding `<CETCompat>false</CETCompat>` to the .csproj of the application opts out of the CET
* Use the Windows Security app, group policy etc. to opt out of the hardware supported stack enforcement for the specific .NET application as described in https://learn.microsoft.com/en-us/defender-endpoint/enable-exploit-protection for details.

### Feature area

Interop

### Affected APIs

_No response_

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #42600.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/9.0.md](https://github.com/dotnet/docs/blob/5ca854ebf098ff6e2041dde7c44a032634174a82/docs/core/compatibility/9.0.md) | [docs/core/compatibility/9.0](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/9.0?branch=pr-en-us-42839) |
| [docs/core/compatibility/interop/9.0/cet-support.md](https://github.com/dotnet/docs/blob/5ca854ebf098ff6e2041dde7c44a032634174a82/docs/core/compatibility/interop/9.0/cet-support.md) | [docs/core/compatibility/interop/9.0/cet-support](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/interop/9.0/cet-support?branch=pr-en-us-42839) |

### CODE REVIEW

File: docs/core/compatibility/interop/9.0/cet-support.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+description: Learn about the breaking change in interop in .NET 8 where CET is supported by default.
+ms.date: 10/04/2024
+---
+# CET supported by default
+
+`apphost` and `singlefilehost` are now marked as Intel CET&mdash;compatible (they're compiled with the `/CETCOMPAT` option). This change was made to enhance security of .NET applications. However, it imposes a limitation on the shared libraries that .NET apps can load and interop with. Libraries aren't allowed to set thread context to a location with an instruction pointer that's not present on the shadow stack or in a table of allowed continuation addresses for exception handling.
+
+## Previous behavior
+
+Previously, shared libraries loaded into the .NET process were able to set thread context using <xref:Microsoft.VisualStudio.CorDebugInterop.ICorDebugProcess.SetThreadContext(System.UInt32,System.UInt32,System.IntPtr)>, `RtlRestoreContext/NtContinue`, or their exception handlers to any location in the process address space.
```
Comment (gewarren) [inline docs/core/compatibility/interop/9.0/cet-support.md:12]: Is this the correct SetThreadContext API?

Comment (janvorli) [reply docs/core/compatibility/interop/9.0/cet-support.md:12]: No, this one is a debugger API. The correct one is https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setthreadcontext.

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Let's :shipit:

File: docs/core/compatibility/interop/9.0/cet-support.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+---
+title: "Breaking change: CET supported by default"
+description: Learn about the breaking change in interop in .NET 8 where CET is supported by default.
```
Comment (janvorli) [inline docs/core/compatibility/interop/9.0/cet-support.md:3]: @gewarren I've just noticed that it mentions .NET 8 here. The change is in .NET 9 though.

Comment (gewarren) [inline docs/core/compatibility/interop/9.0/cet-support.md:3]:
```suggestion
description: Learn about the breaking change in interop in .NET 9 where CET is supported by default.
```

**Pull request merged**